### PR TITLE
[OpenVINO] Fix Mamba 2 Selective SSM representation for Granite 4.0-hybrid

### PR DESCRIPTION
**Pull request by openvino-agent**
# What does this PR do?

Finalize Mamba 2 Selective SSM representation in IR.

Set-up environment:

```sh
pip install -U git+https://github.com/openvino-agent/optimum-intel.git@nemotron_3
```

Export the model:

```sh
optimum-cli export openvino -m ibm-granite/granite-4.0-h-micro granite-4.0-h-micro
optimum-cli export openvino -m ibm-granite/granite-4.0-h-micro granite-4.0-h-micro-int4 --weight-format int4
```

wwb accuracy (similarity) for 10 samples,  `wwb --target-model granite-4.0-h-micro --gt-data granite_gt_fp32.csv --model-type text --num-samples 10`

| model_name                                  | int8      | int4    |
| -------------------------------- | -----    | ------|
| ibm-granite/granite-4.0-h-micro | 0.9713 | 0.8991 |

Inference script:

```python
from transformers import AutoTokenizer, AutoModelForCausalLM
from optimum.intel.openvino import OVModelForCausalLM

#model_path = "ibm-granite/granite-4.0-h-micro"
model_path = "./granite-4.0-h-micro"

tokenizer = AutoTokenizer.from_pretrained(model_path)
model = OVModelForCausalLM.from_pretrained(model_path)

# change input text as desired
input_text = "The capital of France is"
# tokenize the text
input_tokens = tokenizer(input_text, return_tensors="pt")
# generate output tokens
output = model.generate(**input_tokens, max_length=10, do_sample=False)
# decode output tokens into text
output = tokenizer.batch_decode(output)
print(output[0])
```

Fixes # (issue)

## Before submitting
- [N/A] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by regisss** [comment]
> Install transformers==5.10 on top of optimum-intel installations

@echarlaix @rkazants Should we wait for Transformers v5.10 support to be added first before merging this one?

**Review by IlyasMoutawwakil — APPROVED** [review]
LGTM !

File: optimum/exporters/openvino/_ov_ops.py [hunk 1]
```diff
@@ -111,3 +111,158 @@ def convert_recurrent_attention_cell(context):
...
+# The `Mamba2RecurrentCellOp` appears in the Torch graph as a result of replacing the
+# `Mamba2RecurrentCell` `torch.nn.Module` via a registered `ModuleExtension` in the OpenVINO
+# PyTorch frontend; OpenVINO then applies this conversion rule to the resulting operation.
+def convert_recurrent_mamba2_cell(context):
+    # Inputs match the forward signature of `Mamba2RecurrentCell`:
+    #   A          [H]          — negative log-decay rates
+    #   dt         [B, T, H]   — time steps
+    #   B          [B, T, G, N] — input matrix (G groups, expanded to H inside the loop)
+    #   x          [B, T, H, P] — input hidden states
+    #   C          [B, T, G, N] — output matrix (G groups, expanded to H inside the loop)
```
Comment (CuriousPanCake) [inline optimum/exporters/openvino/_ov_ops.py]: Small nit: formally, this happens before the loop body
```suggestion
    #   B          [B, T, G, N] — input matrix (G groups, expanded to H before the loop)
    #   x          [B, T, H, P] — input hidden states
    #   C          [B, T, G, N] — output matrix (G groups, expanded to H before the loop)
```

File: optimum/exporters/openvino/_ov_ops.py [hunk 2]
```diff
@@ -111,3 +111,158 @@ def convert_recurrent_attention_cell(context):
...
+    dt_t = ops.squeeze(dt_t_param, const_one)  # [B, H]
+    B_t = ops.squeeze(B_t_param, const_one)  # [B, H, N]
+    x_t = ops.squeeze(x_t_param, const_one)  # [B, H, P]
+    C_t = ops.squeeze(C_t_param, const_one)  # [B, H, N]
+
+    # Discretization inside the loop body:
+    # dA_t = exp(A * dt_t)  — A is [H], dt_t is [B, H] → broadcast to [B, H]
+    A_unsqueeze = ops.unsqueeze(A, const_zero_axis)  # [1, H]
+    dA_t = ops.exp(ops.multiply(A_unsqueeze, dt_t))  # [B, H]
+    dA_t_4d = ops.unsqueeze(ops.unsqueeze(dA_t, const_minus_one), const_minus_one)  # [B, H, 1, 1]
```
Comment (CuriousPanCake) [inline optimum/exporters/openvino/_ov_ops.py]: Is it possible to use a single Unsqueeze with a list of axis to unsqueeze?

File: optimum/exporters/openvino/_ov_ops.py [hunk 3]
```diff
@@ -111,3 +111,158 @@ def convert_recurrent_attention_cell(context):
...
+    const_minus_one = ops.constant(-1, dtype=np.int32)
+    const_minus_two = ops.constant(-2, dtype=np.int32)
+
+    # Compute heads_per_group = H / G from the shapes of dt and B.
+    dt_shape = ops.shape_of(dt)
+    B_shape = ops.shape_of(B)
+    const_two = ops.constant(2, dtype=np.int32)
+    num_heads = ops.gather(dt_shape, const_two, const_zero_axis)  # H
+    num_groups = ops.gather(B_shape, const_two, const_zero_axis)  # G
+    heads_per_group = ops.convert(ops.divide(num_heads, num_groups), "i64")
```
Comment (CuriousPanCake) [inline optimum/exporters/openvino/_ov_ops.py:165]: I'm not an expert at optimum-intel, but aren't most of dimensions we extract here using ShapeOf + Gather are static models parameters that we can just re-use? This would allow having less ops. in a graph.

File: optimum/exporters/openvino/_ov_ops.py [hunk 4]
```diff
@@ -111,3 +111,158 @@ def convert_recurrent_attention_cell(context):
...
+    B_tiled = ops.tile(B_5d, tile_shape)  # [B, T, G, H/G, N]
+    C_tiled = ops.tile(C_5d, tile_shape)  # [B, T, G, H/G, N]
+
+    # Reshape [B, T, G, H/G, N] → [B, T, H, N]
+    x_shape = ops.shape_of(x)
+    target_4d = ops.gather(x_shape, ops.constant([0, 1], dtype=np.int32), const_zero_axis)
+    N_dim = ops.gather(B_shape, ops.constant(3, dtype=np.int32), const_zero_axis)
+    BC_shape = ops.concat([target_4d, ops.unsqueeze(num_heads, const_zero_axis),
+                           ops.unsqueeze(N_dim, const_zero_axis)], 0)
+    B_expanded = ops.reshape(B_tiled, BC_shape, False)  # [B, T, H, N]
```
Comment (CuriousPanCake) [inline optimum/exporters/openvino/_ov_ops.py]: Instead of constructing `BC_shape`, can we just use OpenVINO's Reshape `special_zero` attribute and reshape into `[0, 0, -1, N]`? So it's all going to be: `B_expanded = ops.reshape(B_tiled, ops.constant([0, 0, -1, N]), True)`

File: optimum/exporters/openvino/_ov_ops.py [hunk 5]
```diff
@@ -111,3 +111,158 @@ def convert_recurrent_attention_cell(context):
...
+    C_expanded = ops.reshape(C_tiled, BC_shape, False)  # [B, T, H, N]
+
+    # Build the zero-initialized output accumulator with shape [B, T, H, P].
+    core_shape = ops.gather(x_shape, ops.constant([0, 1, 2, 3], dtype=np.int32), const_zero_axis)
```
Comment (CuriousPanCake) [inline optimum/exporters/openvino/_ov_ops.py]: Why do we need this Gather if we don't change anything?

File: optimum/exporters/openvino/_ov_ops.py [hunk 6]
```diff
@@ -111,3 +111,163 @@ def convert_recurrent_attention_cell(context):
...
+    last_state_t = ops.parameter([-1, -1, -1, -1], np.float32, "last_state_t")  # [B, H, P, N]
+    core_out_t = ops.parameter([-1, -1, -1, -1], np.float32, "core_out_t")  # [B, T, H, P]
+
+    # Drop the singleton sequence dimension introduced by slicing.
+    dA_t = ops.squeeze(dA_t_param, const_one)  # [B, H]
+    dtB_t = ops.squeeze(dtB_t_param, const_one)  # [B, H, N]
+    x_t = ops.squeeze(x_t_param, const_one)  # [B, H, P]
+    C_t = ops.squeeze(C_t_param, const_one)  # [B, H, N]
+
+    dA_t_4d = ops.unsqueeze(ops.unsqueeze(dA_t, const_minus_one), const_minus_one)  # [B, H, 1, 1]
```
Comment (mryzhov) [inline optimum/exporters/openvino/_ov_ops.py]:
```suggestion
    dA_t_4d = ops.reshape(dA_t_param , ops.constant([0, -1, 1, 1], dtype=np.int64), True)  # [B, H, 1, 1]
```

File: optimum/exporters/openvino/_ov_ops.py [hunk 7]
```diff
@@ -111,3 +111,163 @@ def convert_recurrent_attention_cell(context):
...
+    dA_t_param = ops.parameter([-1, 1, -1], np.float32, "dA_t")  # [B, 1, H]
+    dtB_t_param = ops.parameter([-1, 1, -1, -1], np.float32, "dtB_t")  # [B, 1, H, N]
+    x_t_param = ops.parameter([-1, 1, -1, -1], np.float32, "x_t")  # [B, 1, H, P]
+    C_t_param = ops.parameter([-1, 1, -1, -1], np.float32, "C_t")  # [B, 1, H, N]
+    last_state_t = ops.parameter([-1, -1, -1, -1], np.float32, "last_state_t")  # [B, H, P, N]
+    core_out_t = ops.parameter([-1, -1, -1, -1], np.float32, "core_out_t")  # [B, T, H, P]
+
+    # Drop the singleton sequence dimension introduced by slicing.
+    dA_t = ops.squeeze(dA_t_param, const_one)  # [B, H]
+    dtB_t = ops.squeeze(dtB_t_param, const_one)  # [B, H, N]
```
Comment (mryzhov) [inline optimum/exporters/openvino/_ov_ops.py]:
```suggestion
    dtB_t = ops.reshape(dtB_t_param, ops.constant([0, -1, 1, 0], dtype=np.int64), True)  # [B, H, 1, N]
```

File: optimum/exporters/openvino/_ov_ops.py [hunk 8]
```diff
@@ -111,3 +111,163 @@ def convert_recurrent_attention_cell(context):
...
+    x_t = ops.squeeze(x_t_param, const_one)  # [B, H, P]
+    C_t = ops.squeeze(C_t_param, const_one)  # [B, H, N]
```
Comment (mryzhov) [inline optimum/exporters/openvino/_ov_ops.py]:
```suggestion
    C_t = ops.reshape(C_t_param, ops.constant([0, -1, 1, 0], dtype=np.int64), True)  # [B, H, 1, N]
```

File: optimum/exporters/openvino/_ov_ops.py [hunk 9]
```diff
@@ -111,3 +111,163 @@ def convert_recurrent_attention_cell(context):
...
+
+    dA_t_4d = ops.unsqueeze(ops.unsqueeze(dA_t, const_minus_one), const_minus_one)  # [B, H, 1, 1]
+
+    # dBx_t = dtB_t[:,:,None,:] * x_t[:,:,:,None]  → [B, H, P, N] (outer product, kept in loop)
+    dBx_t = ops.multiply(
+        ops.unsqueeze(dtB_t, const_minus_two),  # [B, H, 1, N]
+        ops.unsqueeze(x_t, const_minus_one),  # [B, H, P, 1]
+    )  # [B, H, P, N]
```
Comment (mryzhov) [inline optimum/exporters/openvino/_ov_ops.py]:
```suggestion
    dBx_t = ops.multiply(dtB_t, x_t)  # [B, H, P, N]
```

File: optimum/exporters/openvino/_ov_ops.py [hunk 10]
```diff
@@ -111,3 +111,163 @@ def convert_recurrent_attention_cell(context):
...
+
+    # state_t = state_{t-1} * dA_t + dBx_t
+    last_state_new = ops.add(ops.multiply(last_state_t, dA_t_4d), dBx_t)  # [B, H, P, N]
```
Comment (mryzhov) [inline optimum/exporters/openvino/_ov_ops.py]: state_updated is better imho
